File: algorithmic-toolbox/src/week1/scala/HugeFibonacci.scala
import scala.io.StdIn._

object HugeFibonacci {
  def main(args: Array[String]) = {
    val in = new java.util.Scanner(System.in)
    val (number, mod) = (in.nextLong, in.nextInt)
    val index = number % getPisano(mod);
    println(fibonacciMod(index.toInt, mod));
  }

  def getPisano(mod: Int): Long = {
    def sumFibonacciMod(start: Int, end: Int, mod: Long) = {
      def sumFibonacciMod(curr: Int, total: Long): Long = {
        if (curr == end) total
        else sumFibonacciMod(curr + 1, total + fibonacciMod(curr, mod))
      }

      sumFibonacciMod(start, 0)
    }

    def getLowerBound(number: Int) = {
      def getLucasNumberIndex(max: Int) = {
        def getLucasNumberIndex(a: Int, b: Int, index: Int): Int = {
          if (b >= max) index
          else getLucasNumberIndex(b, a + b, index + 1)
        }

        getLucasNumberIndex(2, 1, 0);
      }

      2 * getLucasNumberIndex(number)
    }

    def getPisano(ac: Long, c: Long, ap: Long, p: Long, l: Int): Long = {
      if ((ac == ap) && (c == p)) l
      else getPisano(
        fibonacciMod(l, mod),
        c + fibonacciMod(l, mod),
        fibonacciMod(2 * l + 2, mod),
        p - fibonacciMod(l + 1, mod) + fibonacciMod(2 * l + 1, mod) + fibonacciMod(2 * l + 2, mod),
        l + 1)
    }

    val l = getLowerBound(mod)
    getPisano(0, sumFibonacciMod(0, l, mod), 0, sumFibonacciMod(l + 1, 2 * l + 1, mod), l) + 1
  }

  val cache = scala.collection.mutable.Map[Int, Long]()
  def fibonacciMod(n: Int, mod: Long) = {
    def fibonacciMod(index: Int): Long = {
      cache.getOrElseUpdate(index,
        if (index == 0) 0
        else if (index == 1) 1
        else (fibonacciMod(index - 1) + fibonacciMod(index - 2)) % mod);
    }

    fibonacciMod(n)
  }
}

File: algorithmic-toolbox/src/week1/java/HugeFibonacci.java
package week1.java;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

class HugeFibonacci {

    public static void main(final String[] args) {
        try (final Scanner in = new Scanner(System.in)) {
            final long number = in.nextLong();
            final int mod = in.nextInt();

            final int index = (int) (number % getPisano(mod));
            System.out.println(fibonacciMod(index, mod));
        }
    }

    private static int getPisano(final int mod) {
        int length = getLowerBound(mod);

        long current = 0;
        for (int i = 0; i < length; i++) {
            current += fibonacciMod(i, mod);
        }

        long proposed = 0;
        for (int i = length + 1; i <= 2*length; i++) {
            proposed += fibonacciMod(i, mod);
        }

        long appendCurrent = 0, appendProposed = 0;
        while ((proposed != current) || (appendCurrent != appendProposed)) {
            appendCurrent = fibonacciMod(length, mod);
            appendProposed = fibonacciMod(2*length + 2, mod);

            current += appendCurrent;
            proposed -= fibonacciMod(length + 1, mod);
            proposed += fibonacciMod(2*length + 1, mod) + appendProposed;
            length++;
        }

        return length + 1;
    }

    private static final Map<Integer, Long> cache = new HashMap<>();
    private static long fibonacciMod(final int index, final int mod) {
        return cache.computeIfAbsent(index, k -> {
            if (index == 0) {
                return 0L;
            } else if (index == 1) {
                return 1L;
            } else {
                return (fibonacciMod(index - 1, mod) + fibonacciMod(index - 2, mod)) % mod;
            }
        });
    }

    private static int getLowerBound(final int number) {
        return 2*getLucasNumberIndex(number);
    }

    private static int getLucasNumberIndex(final int max) {
        int a = 2, b = 1, index = 1;
        while (b < max) {
            final int tmp = a + b;
            a = b;
            b = tmp;
            index++;
        }

        return (index - 1);
    }

}
